### PR TITLE
Support stdout/stderr file proxy for WASI

### PR DESCRIPTION
**Pull request by kateinoigakukun**
- Implement a Java API to configure a file to be used as the stdout/stderr stream in WASI API.
- Add test cases for WasiConfig.java

### CODE REVIEW

File: wasmtime-jni/src/wasi_config.rs [hunk 1]
```diff
@@ -35,5 +36,16 @@ pub fn ctx_from_java(env: &JNIEnv, obj: JObject) -> Result<WasiCtx> {
         let dir = preopen_dir(&host_path)?;
         builder.preopened_dir(dir, &guest_path);
     }
+
+    if let Some(file) = utils::get_optional_string_field(env, obj, "stdoutFile")? {
+        let file = File::create(file)?;
```
Comment (kateinoigakukun) [inline wasmtime-jni/src/wasi_config.rs:41]: Create a file at the specified path as wasmtime-c-api does.
 https://github.com/bytecodealliance/wasmtime/blob/40d546c420c1e6c4909fc80e048dd1854fda1cd6/crates/c-api/src/wasi.rs#L139-L142

File: wasmtime-jni/src/utils.rs [hunk 2]
```diff
@@ -14,6 +14,19 @@ pub fn get_string_field(env: &JNIEnv, obj: JObject, field: &str) -> Result<Strin
     get_string(env, s)
 }
 
+/// Shorthand of obtaining and turning `JObject` from a field into a Rust `Option<String>`.
+pub fn get_optional_string_field(env: &JNIEnv, obj: JObject, field: &str) -> Result<Option<String>> {
```
Comment (kawamuray) [inline wasmtime-jni/src/utils.rs:18]: For values intended to be exchanged with JNI program, using just a nullable `String` is obviously easier than using `Optional`. Can you change the `WasiConfig` to hold `String` fields and add a function that checks `o.is_null()` and returns `Option<String>` instead?

**Review by kawamuray — CHANGES_REQUESTED** [review]
Thanks for the PR! Left 1 minor comment, PTAL.

File: wasmtime-jni/src/utils.rs (hunk 2 above)
Comment (kateinoigakukun) [reply wasmtime-jni/src/utils.rs]: That makes sense 👍  I changed to use just a nullable `String` for simplexity.

File: src/main/java/io/github/kawamuray/wasmtime/wasi/WasiConfig.java [hunk 3]
```diff
@@ -3,12 +3,15 @@
 import lombok.Builder;
 import lombok.Value;
 import lombok.experimental.Accessors;
+import java.util.Optional;
```
Comment (kawamuray) [inline src/main/java/io/github/kawamuray/wasmtime/wasi/WasiConfig.java:6]: this import is no longer used.

File: src/test/java/io/github/kawamuray/wasmtime/wasi/WasiConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package io.github.kawamuray.wasmtime.wasi;
+import io.github.kawamuray.wasmtime.Store;
+import org.junit.Assert;
+import org.junit.Test;
+import java.util.Optional;
```
Comment (kawamuray) [inline src/test/java/io/github/kawamuray/wasmtime/wasi/WasiConfigTest.java:5]: this import is no longer used.

File: src/test/java/io/github/kawamuray/wasmtime/wasi/WasiConfigTest.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package io.github.kawamuray.wasmtime.wasi;
+import io.github.kawamuray.wasmtime.Store;
+import org.junit.Assert;
```
Comment (kawamuray) [inline src/test/java/io/github/kawamuray/wasmtime/wasi/WasiConfigTest.java:3]: this too apparently?

File: src/main/java/io/github/kawamuray/wasmtime/wasi/WasiConfig.java [hunk 6]
```diff
@@ -19,5 +22,15 @@
     public WasiConfig(String[] args, PreopenDir[] preopenDirs) {
         this.args = args;
         this.preopenDirs = preopenDirs;
+        this.stdoutFile = null;
```
Comment (kawamuray) [inline src/main/java/io/github/kawamuray/wasmtime/wasi/WasiConfig.java:25]: can you make this constructor to call the another one like below to ensure all construction process remain consistent?

```java
this(args, preopenDirs, null, null)
```

**Review by kawamuray — CHANGES_REQUESTED** [review]
Thanks for the follow-up. Just few more points and then LGTM.

**Comment by kateinoigakukun** [comment]
OK, I addressed your review points 👍

**Review by kawamuray — APPROVED** [review]
LGTM, thanks 👍

**Comment by kawamuray** [comment]
Please let me know if you want a new version release for this change.

**Comment by kateinoigakukun** [comment]
I'll send another PR to get exit-code from wasi context in a few days. I'll request you to release a new version after it :)

**Pull request merged**